### PR TITLE
Fix Issue 14097 - root/async.c: free after use

### PR DESCRIPTION
**Pull request by PetarKirov**
This is a patch by Ketmar Dark:
https://issues.dlang.org/show_bug.cgi?id=14097

The fix is to wait for all tasks to finish (as it is [done](https://github.com/D-Programming-Language/dmd/blob/master/src/root/async.c#L227) in the posix version) before calling free().

This is my first commit to the DMD codebase, so I hope I am not doing anything wrong :)

### CODE REVIEW

**Comment by andralex** [comment]
unittest?

**Comment by ZombineDev** [comment]
I am not sure how to test this properly as it is a threading bug. Any hints?

File: src/root/async.c [hunk 1]
```diff
@@ -109,7 +114,8 @@ unsigned __stdcall startthread(void *p)
     AsyncRead *aw = (AsyncRead *)p;
 
     //printf("aw->filesdim = %p %d\n", aw, aw->filesdim);
-    for (size_t i = 0; i < aw->filesdim; i++)
+    size_t dim = aw->filesdim;
+    for (size_t i = 0; i < dim; i++)
```
Comment (WalterBright) [inline src/root/async.c]: Why this change?

Comment (ZombineDev) [inline src/root/async.c]: "... posix [version] `startthread` correctly caches `aw->filesdim`, so it will not use freed memory in loop condition check."

I am not sure if this is needed (I think it should not), but I left it as it is.

**Comment by CyberShadow** [comment]
Ideally, when submitting a patch created by someone else, the `--author` option should be used indicate the real author of the patch. The person who submitted the patch is credited in the git history as the committer.

**Comment by ZombineDev** [comment]
Sorry my mistake. I'll try to fix this.

**Comment by ZombineDev** [comment]
@CyberShadow  I did a `git commit --amend ...` and `git push --force` is it ok now?

**Comment by ZombineDev** [comment]
http://forum.dlang.org/thread/majvk0$a94$86@digitalmars.com#post-mak6ip:24a94:2488:40digitalmars.com

**Comment by CyberShadow** [comment]
Yes, looks good, thanks. Anyway, it was not a problem or blocker, just a nitpick/suggestion.

**Comment by ZombineDev** [comment]
Anyway, a good thing to know in the future ;)

**Comment by ZombineDev** [comment]
I managed to successfully build my branch of dmd, druntime and phobos (master) on Windows, by following the guide at the wiki. However when I run the phobos unittests I get the these linker errors:
https://gist.github.com/ZombineDev/63adab1be1766e92ea21

Probably I have missed some part of the post-build configuration. Has anyone hit this issue before?

**Comment by CyberShadow** [comment]
It can't find the import libraries, or `snn.lib`, the DigitalMars C library. Check the `LIB` setting in `sc.ini`.

**Comment by ZombineDev** [comment]
Thanks for the tip. I fixed the issue, by basically copying the missing *.lib files from the 2.066.1 release to the `%DM_HOME%\dmd2\windows\lib` folder. After a bit of trial and error I got [this][1] output. I'm not sure if it really succeeded in running the unit tests, but I think it did because the on Auto-tester it interprets the same output as a success 

My sc.ini looks like this:
```
[Environment]
LIB="%@P%\..\..\src\phobos;%@P%\..\lib"
DFLAGS="-I%@P%\..\..\src\phobos" "-I%@P%\..\..\src\druntime\import"
LINKCMD=%DM_HOME%\dm\bin\link.exe
```


[1]: https://gist.github.com/ZombineDev/a3fee9b7d1cf521a96d6

**Comment by CyberShadow** [comment]
> After a bit of trial and error I got this output. I'm not sure if it really succeeded in running the unit tests,

Yes, that looks fine.

This line's funny:

> args[0] = '.\unittest.exe¤⌂.\unittest.exe¤⌂♫' 

Culprit:

https://github.com/D-Programming-Language/phobos/blob/master/unittest.d#L92-L94

Why does it do that?

**Comment by ZombineDev** [comment]
Yep, I was wondering the same thing. I think it may be because it was ran through cygwin's make.

File: src/root/async.c (hunk 1 above)
Comment (WalterBright) [inline src/root/async.c]: Ketmar wrote: "'cause `AsyncRead::dispose` can `free()` aw before loop in `startthread` completes, and then `startthread` will try to reference already freed memory in loop condition check. this is very unlikely, but still possible situation. so we have to cache the value."

Please add this as a comment.

**Comment by Safety0ff** [comment]
> Culprit:
> https://github.com/D-Programming-Language/phobos/blob/master/unittest.d#L92-L94
> Why does it do that?

```printf``` expects null terminated string but it is passed a ```char[]``` cast to a ```char*```
It should probably use ```fwrite(args[i].ptr, 1, args[i].length, stdout);``` instead.

File: src/root/async.c (hunk 1 above)
Comment (ZombineDev) [inline src/root/async.c]: Now that `AsyncRead::dispose()` waits for all operations on all files to complete, it should not be possible to `free()` `aw`, before `startthread()` completes.
Even if this somehow happens, then accessing `aw->filesdim` or using a cached version of it is irrelevant, because inside the loop we access the `aw->files` member, which would also lead to the same problem.

Comment (ZombineDev) [inline src/root/async.c]: @WalterBright Because of the above, I think that adding Ketmar's comment would be misleading. Apart from this, the change is a small optimization.

But if you prove me wrong, I would be happy to comply ;)

**Comment by ZombineDev** [comment]
@Safety0ff I think you are correct. However calling `printf` that way shouldn't be a problem as long as there's a null terminator at argv[i][$]. 
The bug may be somewhere in [here][1] but I won't be able to investigate the issue further during this and the next  week, because I'll have exams at my university. Perhaps you guys can look into this in the mean time.

[1]: https://github.com/D-Programming-Language/druntime/blob/dbbea7c5bd4d8936dc8cb523b65f771842f2a487/src/rt/dmain2.d#L306

File: src/root/async.c (hunk 1 above)
Comment (WalterBright) [inline src/root/async.c]: So if that was not the bug, then where is the bug? Does this PR fix anything? :-( I don't think the small optimization does anything.

Comment (CyberShadow) [inline src/root/async.c]: In the first hunk?

Comment (ZombineDev) [inline src/root/async.c]: There are two issues that this PR fixes:

The main thread [calls][1] `aw->start()` (`aw` is an instance of [`AsyncRead`][2]), which kicks off a background thread which starts reading files asynchronously.

The main thread could call `AsyncRead::dispose(aw)` while the background thread is still working (i.e. it is still inside the loop in [`startthread(..)`][3]

The first problem is that before this patch the _WIN32 version of [`AsyncRead::dispose(..)`][4] didn't wait for the background thread to finish and freed the AsyncRead instance, which could crash the background thread inside the [`startthread(..)`][3]. This is what [this][5] change fixes.

The other problem is that it was possible just after the last call to [`SetEvent()`][6] that [`AsyncRead::dispose()`][7] finishes (because all files marked as processed) and frees `aw`, and then the background thread continues and tries to access [`aw->filesdim`][8] for last time and crashes since `aw` is already freed.

My last [comment][9] was incorrect because I didn't see the problem, but Ketmar explained it to me in the [bugzilla issue][10] and now I agree with him.

I hope this makes things clear now.

[1]: https://github.com/D-Programming-Language/dmd/blob/973b7f88c0aa06fcb01334d406544f88aae22238/src/mars.c#L1324

[2]: https://github.com/D-Programming-Language/dmd/blob/f62f211d83cdb7573db6e0d08c54e9d8b9353a8f/src/root/async.h#L23

[3]: https://github.com/D-Programming-Language/dmd/blob/f62f211d83cdb7573db6e0d08c54e9d8b9353a8f/src/root/async.c#L113

[4]: https://github.com/D-Programming-Language/dmd/blob/f62f211d83cdb7573db6e0d08c54e9d8b9353a8f/src/root/async.c#L102

[5]: https://github.com/D-Programming-Language/dmd/pull/4363/files#diff-d15ca38e1ad1fb0a42535fbb6551a62aR102

[6]: https://github.com/D-Programming-Language/dmd/blob/f62f211d83cdb7573db6e0d08c54e9d8b9353a8f/src/root/async.c#L116

[7]: https://github.com/D-Programming-Language/dmd/pull/4363/files#diff-d15ca38e1ad1fb0a42535fbb6551a62aR105

[8]: https://github.com/D-Programming-Language/dmd/blob/f62f211d83cdb7573db6e0d08c54e9d8b9353a8f/src/root/async.c#L112

[9]: https://github.com/D-Programming-Language/dmd/pull/4363#discussion_r24554262

[10]: https://issues.dlang.org/show_bug.cgi?id=14097#c5

Comment (ZombineDev) [inline src/root/async.c]: I will add a comment explaining the problems in code when I have more time (hopefully after Tuesday).

Comment (WalterBright) [inline src/root/async.c]: Yes, please do annotate the code with your great explanation above. This is good stuff, and should be in the code.

**Comment by MartinNowak** [comment]
That "bug" will never manifestate because dispose is only called after synchronizing with each file completion event in read. Maybe there should be an assertion in dispose, that read was called for every file or even simpler asserting that it was called `filedim` times.

Could anyone explain the purpose of spawning a __single__ thread to read in multiple files?
With the added synchronization overhead it seems very unlikely to be faster. Changing the thread priority affects CPU time and shouldn't affect the result.

**Comment by yebblies** [comment]
Does `AsyncRead` actually provide a measurable performance improvement?  Otherwise can we just get rid of it?

**Comment by WalterBright** [comment]
> Does  AsyncRead  actually provide a measurable performance improvement?

I measured it on slow file systems, and yes, it does.

**Comment by yebblies** [comment]
What do yo mean by slow file systems?

**Comment by WalterBright** [comment]
SD cards are slow filesystems, CDs are, USB drives, etc.

**Comment by WalterBright** [comment]
I'd also like to, once dmd is in D, parallelize the lexing and parsing.

**Comment by WalterBright** [comment]
> Could anyone explain the purpose of spawning a single thread to read in multiple files?

Parallelizing file reads doesn't make them faster. But, lexing in parallel with reading files does make it faster.

> Changing the thread priority affects CPU time and shouldn't affect the result.

What that does is cause the file reader to hurry up and block. Then, the lexer runs while the file I/O is blocked. Otherwise, the lexer is running while the file reader has work to be done, and it isn't really any faster.

**Comment by yebblies** [comment]
> SD cards are slow filesystems, CDs are, USB drives, etc.

ie. The least important media types.  I don't see much point in optimizing build speed for anything except SSDs/HDDs.

> I'd also like to, once dmd is in D, parallelize the lexing and parsing.

That sounds much more useful, although we'll need a solution to the shared string table problem.  BTW in ddmd AsyncRead is currently disabled.

**Comment by WalterBright** [comment]
> I don't see much point in optimizing build speed for anything except SSDs/HDDs.

Why remove it?

> BTW in ddmd AsyncRead is currently disabled.

Why? It should be simpler in D.

**Comment by yebblies** [comment]
> Why remove it?

We should be removing complexity that doesn't pull its weight.

> > BTW in ddmd AsyncRead is currently disabled.

> Why? It should be simpler in D.

It didn't autoconvert and I didn't bother porting it manually.  It's probably not hard to do, but why bother if it doesn't improve performance?

File: src/root/async.c [hunk 2]
```diff
@@ -99,6 +99,11 @@ int AsyncRead::read(size_t i)
 
 void AsyncRead::dispose(AsyncRead *aw)
 {
+    for (size_t i = 0; i < aw->filesdim; i++)
+    {
+        FileData *f = &aw->files[i];
+        WaitForSingleObject(f->event, INFINITE);
+    }
     free(aw);
```
Comment (MartinNowak) [inline src/root/async.c]: Can we remove this part of the change? It's asymmetric to the POSIX version and there is no need to call dispose while still reading. You could make it an assertion that read was called for each file.

**Comment by ibuclaw** [comment]
> We should be removing complexity that doesn't pull its weight.

GDC disables AsyncRead because of cross-platform complications in it.  But IMO only targeting SSD/HDDs is a bad choice in todays market, eg: ARM. http://www.tinygreenpc.com/computers/media-players/minix-neo-x8-h-plus.html

**Comment by CyberShadow** [comment]
Um, this is in the compiler, and you'll still be doing *development* work on a real computer.

**Comment by ibuclaw** [comment]
ARM is a real computer.

**Comment by ibuclaw** [comment]
And native compilers exist *today* for *many* server and embedded targets: https://packages.debian.org/sid/gdc-5

**Comment by CyberShadow** [comment]
AsyncRead should be rewritten to:

- use non-blocking I/O instead of threads (this should also fix portability issues)
- read as many files at once as possible (maximize IO queue depth)

**Comment by CyberShadow** [comment]
> And native compilers exist today for many server and embedded targets: https://packages.debian.org/sid/gdc-5

Yes, but just because you can now use a native compiler instead of cross-compiling, that doesn't mean that you'll be doing the *bulk of development work* on them.

**Comment by ibuclaw** [comment]
> Yes, but just because you can now use a native compiler instead of cross-compiling, that doesn't mean that you'll be doing the bulk of development work on them.

Build farms, such as those used by Linux distributions will benefit the most, for sure.  (One day, it would be nice to get ARM working with the Autotester for GDC/LDC PRs).

**Comment by yebblies** [comment]
We're talking about DMD, which doesn't even target ARM.  DMD's host platform is probably always going to be a desktop or a server with a HDD or SSD.

Hopefully in the future we'll refactor (D)DMD with parallelism in mind, and this will be fairly straightforward to implement.  But as I said - I haven't implemented `AsyncRead` for DDMD, and I'm not really planning to unless it's worth doing.

**Comment by WalterBright** [comment]
> •use non-blocking I/O instead of threads (this should also fix portability issues)

This is a simple thread. Why should that not be portable?

> •read as many files at once as possible (maximize IO queue depth)

I/O is physically not capable of reading multiple files at once, so this is pointless. It works fine to have one thread reading the files sequentially, and the main thread parsing the previous file while the next file is being read. It doesn't need to be cleverer than this.

BTW, this is how optlink achieves its speed.

**Comment by CyberShadow** [comment]
> This is a simple thread. Why should that not be portable?

Non-blocking file descriptors are part of UNIX and work without linking to additional libraries. Not so with threads.

> I/O is physically not capable of reading multiple files at once, so this is pointless.

Totally wrong!

1. DMD doesn't and can't know where the files are on the physical disk. For platter drives, issuing as many reading commands at once allows the OS/controller to reorder the reads in sequential order, so the drive head has to do as little reading as possible.
2. If the files are on different physical drives (e.g. they're part of a cluster or RAID set), they CAN be read simultaneously.
3. Solid-state memory *can* read from multiple locations at once!

SATA has a queue depth of 32 operations! It is stupid to not take advantage of it.

**Comment by MartinNowak** [comment]
>> This is a simple thread. Why should that not be portable?

> It's less portable. Non-blocking file descriptors are part of UNIX and work without linking to additional libraries. Not so with threads.

There is no portable async file solution for UNIX, non-blocking file descriptors only work for network file systems.

> > I/O is physically not capable of reading multiple files at once, so this is pointless.

>>Totally wrong!

Yep, fastest way to read many files is to issue as many read commands as possible, the kernel will schedule disc access.

This could be very nicely modeled using async and futures.

Let's get back to the review though, it fixes a latent bug, but I don't agree with adding the WaitForSingleObject loop in dispose.

File: src/root/async.c (hunk 2 above)
Comment (MartinNowak) [inline src/root/async.c]: @ZombineDev can you remove that part?

**Comment by MartinNowak** [comment]
Ping. Change the patch and fix the bug?

**Comment by ZombineDev** [comment]
#4653

File: src/root/async.c (hunk 2 above)
Comment (ZombineDev) [inline src/root/async.c]: Done in #4653.

**Comment by ZombineDev** [comment]
IMO, the current API (of `AsyncRead`) is quite error-prone. It would be beneficial if a more general strategy for parallelism is employed throughout the compiler (not only async file reading, but for all tasks that can be processed independently) such as, for example, future/promises.
After the frontend development transitions from C++ to D I would like to help with this.

**Comment by MartinNowak** [comment]
> After the frontend development transitions from C++ to D I would like to help with this.

There is quite some work necessary to make this work, e.g. making the backend thread safe.
A future/promise based solution backed by some worker threads should indeed work nicely.

**Comment by MartinNowak** [comment]
[SDC](ttps://github.com/deadalnix/SDC) gets this right.
